Index: libs/xmlrpc-c/lib/abyss/src/socket_unix.c
===================================================================
--- libs/xmlrpc-c/lib/abyss/src/socket_unix.c	(revision 13903)
+++ libs/xmlrpc-c/lib/abyss/src/socket_unix.c	(working copy)
@@ -227,20 +227,8 @@
     struct socketUnix * const socketUnixP = channelP->implP;
 
     int rc;
-	fd_set rfds, efds;
-	struct timeval tv = { 10, 0 };
-	
-	FD_ZERO(&rfds);
-	FD_ZERO(&efds);
-	FD_SET(socketUnixP->fd, &rfds);
-	FD_SET(socketUnixP->fd, &efds);
+    rc = recv(socketUnixP->fd, buffer, bufferSize, 0);
 
-	if ((rc = select(socketUnixP->fd + 1, &rfds, NULL, &efds, &tv)) > 0) {
-		rc = recv(socketUnixP->fd, buffer, bufferSize, 0);
-	} else {
-		rc = -1;
-	}
-
     if (rc < 0) {
         *failedP = TRUE;
         if (ChannelTraceIsActive)
