Index: mod_skypiax.c
===================================================================
--- mod_skypiax.c	(revision 16576)
+++ mod_skypiax.c	(working copy)
@@ -1741,7 +1741,7 @@
 
 	if (channel) {
 
-		if (!switch_channel_test_flag(channel, CF_BRIDGED)) {
+		//FIXME if (!switch_channel_test_flag(channel, CF_BRIDGED)) {
 
 			switch_dtmf_t dtmf = { (char) value[0], switch_core_default_dtmf_duration(0) };
 			DEBUGA_SKYPE("received DTMF %c on channel %s\n", SKYPIAX_P_LOG, dtmf.digit, switch_channel_get_name(channel));
@@ -1750,10 +1750,10 @@
 			switch_channel_queue_dtmf(channel, &dtmf);
 			switch_set_flag(tech_pvt, TFLAG_DTMF);
 			switch_mutex_unlock(tech_pvt->flag_mutex);
-		} else {
-			NOTICA
-				("received a DTMF on channel %s, but we're BRIDGED, so let's NOT relay it out of band\n", SKYPIAX_P_LOG, switch_channel_get_name(channel));
-		}
+		//FIXME } else {
+			//FIXME NOTICA
+				//FIXME ("received a DTMF on channel %s, but we're BRIDGED, so let's NOT relay it out of band\n", SKYPIAX_P_LOG, switch_channel_get_name(channel));
+		//FIXME }
 	} else {
 		WARNINGA("received %c DTMF, but no channel?\n", SKYPIAX_P_LOG, value[0]);
 	}
Index: skypiax_protocol.c
===================================================================
--- skypiax_protocol.c	(revision 16576)
+++ skypiax_protocol.c	(working copy)
@@ -96,7 +96,7 @@
 #ifdef WIN32
 	sockbufsize = SAMPLES_PER_FRAME * 8 * 3;
 #else
-	sockbufsize = SAMPLES_PER_FRAME * 8;
+	sockbufsize = SAMPLES_PER_FRAME * 8 * 3;
 #endif //WIN32
 	size = sizeof(int);
 	setsockopt(s, SOL_SOCKET, SO_RCVBUF, (char *) &sockbufsize, size);
@@ -113,7 +113,7 @@
 #ifdef WIN32
 	sockbufsize = SAMPLES_PER_FRAME * 8 * 3;
 #else
-	sockbufsize = SAMPLES_PER_FRAME * 8;
+	sockbufsize = SAMPLES_PER_FRAME * 8 * 3;
 #endif //WIN32
 	size = sizeof(int);
 	setsockopt(s, SOL_SOCKET, SO_SNDBUF, (char *) &sockbufsize, size);
