--- thread.cpp.bak	2005-07-19 08:16:10.000000000 -0400
+++ thread.cpp	2005-11-14 13:04:59.200854500 -0500
@@ -1758,14 +1758,16 @@
 int SysTime::getTimeOfDay(struct timeval *tp) 
 {
 	struct timeval temp;
-    int ret(0);
+	time_t now;
+	int ret(0);
     lock();
 
 #ifdef	WIN32
     // We could use _ftime(), but it is not available on WinCE.
     // (WinCE also lacks time.h)
     // Note also that the average error of _ftime is around 20 ms :)
-    time(&temp.tv_sec);
+	time(&now);
+	temp.tv_sec = (long) now;
     temp.tv_usec = (GetTickCount() % 1000) * 1000;
 	memcpy(tp, &temp, sizeof(struct timeval));
 #else
--- lockfile.cpp.bak	2005-04-23 19:08:21.000000000 -0400
+++ lockfile.cpp	2005-11-14 13:05:56.305244100 -0500
@@ -68,7 +68,7 @@
 bool Lockfile::lock(const char *name)
 {
 	char mname[65];
-	char *ext = strrchr(name, '/');
+	char *ext = (char *) strrchr(name, '/');
 	
 	if(ext)
 		name = ++ext;
--- socket.cpp.bak	2005-10-09 09:12:35.000000000 -0400
+++ socket.cpp	2005-11-14 13:05:32.710375300 -0500
@@ -1330,7 +1330,7 @@
 		return;
 	}
 
-#if defined(SO_REUSEADDR) && !defined(WIN32)
+#if defined(SO_REUSEADDR) //&& !defined(WIN32)
         int opt = 1;
         setsockopt(so, SOL_SOCKET, SO_REUSEADDR, (char *)&opt,
                 (socklen_t)sizeof(opt));
