diff --git a/Makefile b/Makefile
index 8f21f7c..79e7a07 100644
--- a/Makefile
+++ b/Makefile
@@ -233,7 +233,7 @@ NACL_ARCHES = nacl_ia32 nacl_x64
 # List of files that trigger Makefile regeneration:
 GYPFILES = build/all.gyp build/features.gypi build/standalone.gypi \
            build/toolchain.gypi samples/samples.gyp src/d8.gyp \
-           test/cctest/cctest.gyp tools/gyp/v8.gyp
+           tools/gyp/v8.gyp
 
 # If vtunejit=on, the v8vtune.gyp will be appended.
 ifeq ($(vtunejit), on)
diff --git a/build/all.gyp b/build/all.gyp
index 5fbd8c2..9c2a4db 100644
--- a/build/all.gyp
+++ b/build/all.gyp
@@ -10,7 +10,6 @@
       'dependencies': [
         '../samples/samples.gyp:*',
         '../src/d8.gyp:d8',
-        '../test/cctest/cctest.gyp:*',
       ],
       'conditions': [
         ['component!="shared_library"', {
