--- export/src/editline/readline.h	2007-03-02 22:26:59.000000000 +0100
+++ ../src/editline/readline.h	2007-03-02 22:38:45.000000000 +0100
@@ -45,9 +45,11 @@
 typedef char	 *CPFunction(const char *, int);
 typedef char	**CPPFunction(const char *, int, int);
 
+typedef void *histdata_t;
+
 typedef struct _hist_entry {
 	const char	*line;
-	const char	*data;
+	histdata_t	*data;
 } HIST_ENTRY;
 
 typedef struct _keymap_entry {
@@ -68,7 +70,7 @@
 
 #ifndef CTRL
 #include <sys/ioctl.h>
-#if !defined(__sun__) && !defined(__hpux__)
+#ifdef __GLIBC__
 #include <sys/ttydefaults.h>
 #endif
 #ifndef CTRL
