--- export/src/unvis.c	2007-03-02 22:26:59.000000000 +0100
+++ ../src/unvis.c	2007-03-02 22:38:45.000000000 +0100
@@ -29,7 +29,7 @@
  * SUCH DAMAGE.
  */
 
-#include <sys/cdefs.h>
+#include <config.h>
 #if defined(LIBC_SCCS) && !defined(lint)
 #if 0
 static char sccsid[] = "@(#)unvis.c	8.1 (Berkeley) 6/4/93";
@@ -38,7 +38,6 @@
 #endif
 #endif /* LIBC_SCCS and not lint */
 
-#include "namespace.h"
 #include <sys/types.h>
 
 #include <assert.h>
