--- export/src/vis.h	2007-08-31 00:02:46.000000000 +0200
+++ ../src/vis.h	2007-08-31 00:03:08.000000000 +0200
@@ -34,7 +34,7 @@
 #ifndef _VIS_H_
 #define	_VIS_H_
 
-#include <sys/types.h>
+#include <config.h>
 
 /*
  * to select alternate encoding format
@@ -72,8 +72,6 @@
  */
 #define	UNVIS_END	1	/* no more characters */
 
-#include <sys/cdefs.h>
-
 __BEGIN_DECLS
 char	*vis(char *, int, int, int);
 char	*svis(char *, int, int, int, const char *);
@@ -84,7 +82,7 @@
 int	strunvis(char *, const char *);
 int	strunvisx(char *, const char *, int);
 #ifndef __LIBC12_SOURCE__
-int	unvis(char *, int, int *, int) __RENAME(__unvis13);
+int	unvis(char *, int, int *, int);
 #endif
 __END_DECLS
 
