--- export/src/el_term.h	2007-08-31 00:02:46.000000000 +0200
+++ ../src/el_term.h	2007-08-31 00:03:08.000000000 +0200
@@ -81,6 +81,15 @@
 #define	A_K_EN		5
 #define	A_K_NKEYS	6
 
+#ifdef _SUNOS
+extern int tgetent(char *, const char *);
+extern int tgetflag(char *);
+extern int tgetnum(char *);
+extern int tputs(const char *, int, int (*)(int));
+extern char* tgoto(const char*, int, int);
+extern char* tgetstr(char*, char**);
+#endif
+
 protected void	term_move_to_line(EditLine *, int);
 protected void	term_move_to_char(EditLine *, int);
 protected void	term_clear_EOL(EditLine *, int);
