--- export/src/editline/readline.h	2007-08-31 00:02:46.000000000 +0200
+++ ../src/editline/readline.h	2007-08-31 00:03:08.000000000 +0200
@@ -46,9 +46,11 @@
 typedef char	**CPPFunction(const char *, int, int);
 typedef char     *rl_compentry_func_t(const char *, int);
 
+typedef void *histdata_t;
+
 typedef struct _hist_entry {
 	const char	*line;
-	const char	*data;
+	histdata_t	*data;
 } HIST_ENTRY;
 
 typedef struct _keymap_entry {
@@ -69,7 +71,7 @@
 
 #ifndef CTRL
 #include <sys/ioctl.h>
-#if !defined(__sun__) && !defined(__hpux__)
+#ifdef __GLIBC__
 #include <sys/ttydefaults.h>
 #endif
 #ifndef CTRL
