--- export/examples/tc1.c	2007-08-31 00:02:46.000000000 +0200
+++ ../examples/tc1.c	2007-08-31 00:03:08.000000000 +0200
@@ -32,20 +32,6 @@
  * SUCH DAMAGE.
  */
 
-#include "config.h"
-#ifndef lint
-__COPYRIGHT("@(#) Copyright (c) 1992, 1993\n\
-	The Regents of the University of California.  All rights reserved.\n");
-#endif /* not lint */
-
-#if !defined(lint) && !defined(SCCSID)
-#if 0
-static char sccsid[] = "@(#)test.c	8.1 (Berkeley) 6/4/93";
-#else
-__RCSID("$NetBSD: tc1.c,v 1.1 2006/08/31 20:20:38 rpaulo Exp $");
-#endif
-#endif /* not lint && not SCCSID */
-
 /*
  * test.c: A little test program
  */
@@ -60,6 +46,11 @@
 
 #include "histedit.h"
 
+/* from src/sys/sys/cdefs.h */
+#ifndef __UNCONST
+# define __UNCONST(a) ((void *)(unsigned long)(const void *)(a))
+#endif
+
 static int continuation = 0;
 volatile sig_atomic_t gotsig = 0;
 
