--- export/doc/editrc.5.roff	2007-08-31 00:02:46.000000000 +0200
+++ ../doc/editrc.5.roff	2007-08-31 00:03:08.000000000 +0200
@@ -157,7 +157,7 @@
 .Ar command
 can contain control characters of the form
 .Sm off
-.Sq No ^ Ar character
+.Sq ^character
 .Sm on
 .Po
 e.g.
@@ -183,7 +183,7 @@
 .It Ic \ev
 Vertical tab
 .Sm off
-.It Sy \e Ar nnn
+.It Sy \ennn
 .Sm on
 The ASCII character corresponding to the octal number
 .Ar nnn .
@@ -222,11 +222,11 @@
 causing an error.
 .Fl v
 causes messages to be verbose.
-.It Ic edit Op Li on | Li off
+.It Ic edit Op on | off
 Enable or disable the
 .Nm editline
 functionality in a program.
-.It Ic history Ar list | Ar size Dv n | Ar unique Dv n
+.It Ic history Ar list | Ar size n | Ar unique n
 The 
 .Ar list
 command lists all entries in the history.
